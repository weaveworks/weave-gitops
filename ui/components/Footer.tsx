import * as React from "react";
import styled from "styled-components";
import p from "../../package.json";
import { useVersion } from "../hooks/version";
import { GetVersionResponse } from "../lib/api/core/core.pb";
import { getAppVersion } from "../lib/utils";
import Flex from "./Flex";
import Link from "./Link";
import Spacer from "./Spacer";
import Text from "./Text";
import Version, { VersionProps } from "./Version";

type Props = {
  className?: string;
};

const RightFoot = styled(Flex)`
  padding-right: ${(props) => props.theme.spacing.medium};
`;

const LeftFoot = styled(Flex)``;

function Footer({ className }: Props) {
  const { data, isLoading } = useVersion();
  const versionData = data || ({} as GetVersionResponse);

  const appVersion = getAppVersion(data, p.version, isLoading, "v");

  const versions: VersionProps[] = !isLoading
    ? [
        {
          productName: "Kubernetes",
          appVersion: {
            versionText: versionData.kubeVersion,
          },
        },
        {
          productName: "Weave GitOps",
          appVersion: appVersion,
        },
      ]
    : [];

  return (
    <Flex as="footer" wide between className={className} role="footer">
      <LeftFoot>
        <Text noWrap>Need help? Contact us at</Text>
        <Spacer padding="xxs" />
        <Link newTab href="mailto:info@weaveworks.org">
          info@weaveworks.org
        </Link>
      </LeftFoot>
      <RightFoot>
        {!isLoading &&
          versions.map((version, index) => (
            <React.Fragment key={`version ${index}`}>
              <Spacer padding="xxs" />
              <Version {...version} />
            </React.Fragment>
          ))}
        <Spacer padding="xxs" />
        <Text noWrap>Â© {new Date().getFullYear()} Weaveworks</Text>
      </RightFoot>
    </Flex>
  );
}

export default styled(Footer).attrs({ className: Footer.name })`
  max-width: 100%;
  box-sizing: border-box;
  background: ${(props) => props.theme.colors.neutralGray};
  color: ${(props) => props.theme.colors.neutral30};
  padding: ${(props) => props.theme.spacing.medium};
  border-radius: 0 0
    ${(props) => `${props.theme.spacing.xs} ${props.theme.spacing.xs}`};
`;
