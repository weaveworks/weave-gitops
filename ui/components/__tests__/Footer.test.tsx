import { screen, waitFor } from "@testing-library/dom";
import { render } from "@testing-library/react";
import "jest-canvas-mock";
import "jest-styled-components";
import React, { act } from "react";
import {
  createCoreMockClient,
  withContext,
  withTheme,
} from "../../lib/test-utils";
import Footer from "../Footer";

jest.mock(
  "../../../package.json",
  () => ({
    version: "x.y.z",
  }),
  { virtual: true },
);

describe("Footer", () => {
  let container;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  describe("snapshots", () => {
    it("default", async () => {
      await act(async () => {
        render(
          withTheme(
            withContext(<Footer />, "/", {
              api: createCoreMockClient({
                GetVersion: () => ({
                  semver: "v0.0.1",
                  branch: "mybranch",
                  commit: "123abcd",
                }),
              }),
            }),
          ),
          container,
        );
      });

      await waitFor(() => expect(screen.getByText("Weave GitOps:")));
      const footer = screen.getByRole("footer");
      expect(footer).toMatchSnapshot();
    });
    it("no api version", async () => {
      await act(async () => {
        render(
          withTheme(
            withContext(<Footer />, "/", {
              api: createCoreMockClient({
                GetVersion: () => ({}),
              }),
            }),
          ),
          container,
        );
      });

      await waitFor(() => expect(screen.getByText("Weave GitOps:")));
      const footer = screen.getByRole("footer");
      expect(footer).toMatchSnapshot();
    });
  });
});
