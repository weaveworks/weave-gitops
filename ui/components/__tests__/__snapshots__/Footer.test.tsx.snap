// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer snapshots default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c3 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  white-space: nowrap;
}

.c6 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  color: #00b3ec;
}

.c8 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 600;
  text-transform: none;
  font-style: normal;
  white-space: nowrap;
}

.c9 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
}

.c10 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 600;
  text-transform: none;
  font-style: normal;
}

.c4 {
  padding: 4px;
}

.c5 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7 {
  padding-right: 24px;
}

.c1 {
  max-width: 100%;
  box-sizing: border-box;
  background: #F6F7F9;
  color: #737373;
  padding: 24px;
  border-radius: 0 0 8px 8px;
}

<footer
  class="c0 c1 Footer"
  role="footer"
>
  <div
    class="c2"
  >
    <span
      class="c3"
    >
      Need help? Contact us at
    </span>
    <div
      class="c4"
    />
    <a
      class="c5 Link"
      href="mailto:sales@weave.works"
      rel="noreferrer"
      target="_blank"
    >
      <span
        class="c6"
        color="primary"
      >
        sales@weave.works
      </span>
    </a>
  </div>
  <div
    class="c2 c7"
  >
    <div
      class="c4"
    />
    <div
      class="c2 Version"
    >
      <span
        class="c8"
      >
        Kubernetes
        :
      </span>
      <div
        class="c4"
      />
      <span
        class="c9"
      >
        -
      </span>
    </div>
    <div
      class="c4"
    />
    <div
      class="c2 Version"
    >
      <span
        class="c8"
      >
        Weave GitOps
        :
      </span>
      <div
        class="c4"
      />
      <a
        class="c5 Link"
        href="https://github.com/weaveworks/weave-gitops/commit/123abcd"
        rel="noreferrer"
        target="_blank"
      >
        <span
          class="c6"
          color="primary"
        >
          <span
            class="c10"
          >
            mybranch-123abcd
          </span>
        </span>
      </a>
    </div>
    <div
      class="c4"
    />
    <span
      class="c3"
    >
      © 
      2024
       Weaveworks
    </span>
  </div>
</footer>
`;

exports[`Footer snapshots no api version 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c3 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  white-space: nowrap;
}

.c6 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  color: #00b3ec;
}

.c8 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 600;
  text-transform: none;
  font-style: normal;
  white-space: nowrap;
}

.c9 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
}

.c10 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 600;
  text-transform: none;
  font-style: normal;
}

.c4 {
  padding: 4px;
}

.c5 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7 {
  padding-right: 24px;
}

.c1 {
  max-width: 100%;
  box-sizing: border-box;
  background: #F6F7F9;
  color: #737373;
  padding: 24px;
  border-radius: 0 0 8px 8px;
}

<footer
  class="c0 c1 Footer"
  role="footer"
>
  <div
    class="c2"
  >
    <span
      class="c3"
    >
      Need help? Contact us at
    </span>
    <div
      class="c4"
    />
    <a
      class="c5 Link"
      href="mailto:sales@weave.works"
      rel="noreferrer"
      target="_blank"
    >
      <span
        class="c6"
        color="primary"
      >
        sales@weave.works
      </span>
    </a>
  </div>
  <div
    class="c2 c7"
  >
    <div
      class="c4"
    />
    <div
      class="c2 Version"
    >
      <span
        class="c8"
      >
        Kubernetes
        :
      </span>
      <div
        class="c4"
      />
      <span
        class="c9"
      >
        -
      </span>
    </div>
    <div
      class="c4"
    />
    <div
      class="c2 Version"
    >
      <span
        class="c8"
      >
        Weave GitOps
        :
      </span>
      <div
        class="c4"
      />
      <a
        class="c5 Link"
        href="https://github.com/weaveworks/weave-gitops/releases/tag/v0.38.0"
        rel="noreferrer"
        target="_blank"
      >
        <span
          class="c6"
          color="primary"
        >
          <span
            class="c10"
          >
            v0.38.0
          </span>
        </span>
      </a>
    </div>
    <div
      class="c4"
    />
    <span
      class="c3"
    >
      © 
      2024
       Weaveworks
    </span>
  </div>
</footer>
`;
